
#### KvarQ: Targeted and direct variant calling in FastQ reads of bacterial genomes

The program KvarQ performs rapid in silico genotyping for selected loci (e.g.
phylogenetic SNPs, drug resistance mutations) in bacterial genome sequences in
[FastQ] format. Mapping to a whole-genome reference sequence or de novo
assembly or the short reads is not necessary.

Some example testsuites can be found in the `testsuites/` subdirectory.

KvarQ is written in Python and C, is open-source and published under [GNU GPLv3].

Pre-compiled packages are available for Windows 7 or later, and for Mac OS X
(10.6 or later) can be downloaded from the [Swiss TPH website].


##### Documentation

For more information refer to the [online documentation].


##### [How to cite]

> KvarQ: targeted and direct variant calling from fastq reads of bacterial genomes.
> Steiner A, Stucki D, Coscolla M, Borrell S, Gagneux S.
> BMC Genomics. 2014 Oct 9;15(1):881. doi: 10.1186/1471-2164-15-881.
> <http://www.swisstph.ch/kvarq>.


[FastQ]: <https://en.wikipedia.org/wiki/FASTQ_format>
[MTBC]: <https://en.wikipedia.org/wiki/MTBC>
[Swiss TPH website]: <http://www.swisstph.ch/kvarq>
[GNU GPLv3]: <http://www.gnu.org/licenses/gpl-3.0.html>
[online documentation]: <http://kvarq.readthedocs.org>
[How to cite]: <http://kvarq.readthedocs.org/en/latest/cite.html>
